lib/semantics/io.h

```diff
@@ -0,0 +1,145 @@
+  void Enter(const parser::IdVariable &);
+  void Enter(const parser::InputItem &);
+  void Enter(const parser::InquireSpec &);
+  void Enter(const parser::InquireSpec::CharVar &);
+  void Enter(const parser::InquireSpec::IntVar &);
+  void Enter(const parser::InquireSpec::LogVar &);
+  void Enter(const parser::IoControlSpec &);
+  void Enter(const parser::IoControlSpec::Asynchronous &);
+  void Enter(const parser::IoControlSpec::CharExpr &);
+  void Enter(const parser::IoControlSpec::Pos &);
+  void Enter(const parser::IoControlSpec::Rec &);
+  void Enter(const parser::IoControlSpec::Size &);
+  void Enter(const parser::IoUnit &);
+  void Enter(const parser::MsgVariable &);
+  void Enter(const parser::OutputItem &);
+  void Enter(const parser::StatusExpr &);
+  void Enter(const parser::StatVariable &);
+
+  void Leave(const parser::BackspaceStmt &);
+  void Leave(const parser::CloseStmt &);
+  void Leave(const parser::EndfileStmt &);
+  void Leave(const parser::FlushStmt &);
+  void Leave(const parser::InquireStmt &);
+  void Leave(const parser::OpenStmt &);
+  void Leave(const parser::ReadStmt &);
+  void Leave(const parser::RewindStmt &);
+  void Leave(const parser::WaitStmt &);
+  void Leave(const parser::WriteStmt &);
+
+private:
+  // Nothing to check for Print, but include it anyway.
+  ENUM_CLASS(IoStmtKind, None, Backspace, Close, Endfile, Flush, Inquire, Open,
+      Print, Read, Rewind, Wait, Write);
+
+  // Union of specifiers for all statements.
+  ENUM_CLASS(SpecifierKind, Access, Action, Advance, Asynchronous, Blank,
+      Decimal, Delim, Direct, Encoding, End, Eor, Err, Exist, File, Fmt, Form,
+      Formatted, Id, Iomsg, Iostat, Name, Named, Newunit, Nextrec, Nml, Number,
+      Opened, Pad, Pending, Pos, Position, Read, Readwrite, Rec, Recl, Round,
+      Sequential, Sign, Size, Status, Stream, Unformatted, Unit, Write,
+      Convert,  // nonstandard
+      Dispose,  // nonstandard
+  );
+
+  void LeaveReadWrite() const;
+
+  void SetSpecifier(SpecifierKind);
+
+  void CheckStringValue(
+      SpecifierKind, const std::string &, const parser::CharBlock &) const;
+
+  void CheckForRequiredSpecifier(SpecifierKind) const;
+  void CheckForRequiredSpecifier(bool, const std::string &) const;
+  void CheckForRequiredSpecifier(SpecifierKind, SpecifierKind) const;
+  void CheckForRequiredSpecifier(
+      SpecifierKind, bool, const std::string &) const;
+  void CheckForRequiredSpecifier(
+      bool, const std::string &, SpecifierKind) const;
+  void CheckForRequiredSpecifier(
+      bool, const std::string &, bool, const std::string &) const;
+
+  void CheckForProhibitedSpecifier(SpecifierKind) const;
+  void CheckForProhibitedSpecifier(SpecifierKind, SpecifierKind) const;
+  void CheckForProhibitedSpecifier(
+      SpecifierKind, bool, const std::string &) const;
+  void CheckForProhibitedSpecifier(
+      bool, const std::string &, SpecifierKind) const;
+
+  void Init(IoStmtKind s) {
+    stmt_ = s;
```

Comment:
You might want to use some means to make your state data members more future-proof; right now, if your `Init()` member function were to omit initialization of some data member, it would still compile and the bug would be hard to find.

Reply:
I've packaged all independent class flags together in a Flag enum set.  That makes initialization automatic.  That also makes uses of the flags more in character with those for specifiers.